### ISSUE TITLE
creating and accessing C# objects

### ISSUE THREAD

**Issue by hanlectin** [FlorianRappl/Mages#112]
Hi Florian!

I'm probably missing something, or maybe it's not (yet) possible at all? 
-> I would like to create a C# object (in mages), and then access it (from mages).

what I do now: I have a plugin-function called "object" [public static object @object(Object[] args)], which I call to create/instanitante a custom C# object (it's internally added to a dictionary of counters, so I can later refer to it again, see below...):
`   object("counter1", "new:counter", 0, 100, 100, 5);`
which creates a (named) counter "counter1" with some properties

I can later use
   ```
object("counter1", "start");
   object("counter1", "stop");
   object("counter1", "setvalue", 15);
```
   etc.
to work with this named counter.

however, it would be great to use something like:
```
  counter1 = new counter(0,100,100,5);
  counter1.start();
  counter1.stop();
  counter1.value = 15;
```

[Object metadata #49](https://github.com/FlorianRappl/Mages/issues/49) might be what I'm looking for. It was added to milestone 2.0 - but is still open... 

Thanks for any info/hints/pointers!

**Comment by hanlectin** [comment]
thanks for the quick reply!

I already tried that. In fact, it was my first approach, as I thought I found in the github-documentation and -issues, that it's supposed to work... but it didn't work :(

...now trying to send you a simplified snipped of my code, I (as so often when doing this) found the problem:
I do have a little "preprocessor" on my scripts replacing '.' (dots) with "_o_" -> so 
`counter1.start();`
becomes
`counter1_o_start();`
which of course doesn't work :/ (stupid me)

reason for this: in our application we use hirarchical variable notation, i.e. 
`root.module1.somevar = 2;`

as the '.' is not part of a valid identifier, I always perform the above replacement for non-numeric groups outside quotes
(and vice-versa when showing the variables)

so, Mages works perfectly (as always)...

--> seems I need to find a more sophisticated way of replacing the dots... unless you've got an idea on how to allow dots as part of an identifier without messing with the core functionality

thank you!

**Comment by hanlectin** [comment]
hmmm... if that would really be an option int the foreseeable future: GREAT!

the current definition of Identifiers is as follows:
```
name_start_character ::= [#x80 - #xFFFF] | letter | "_"
name_character ::= name_start_character | digit
identifier ::= name_start_character name_character*
```
changing this to:
```
name_start_character ::= [#x80 - #xFFFF] | letter | "_"
name_character ::= name_start_character | digit | '.'
identifier ::= name_start_character name_character*
```
would most likely do the trick for me - however, wouldn't that result in ambiguities? - like:
let's say I have a term: "root.module1.counter1.value = 42" (where counter1 is NOT an instance of the counter object)
or having e.g. a counter object called "root.module1.counter1" (which has a "value" property)
--> the mages-parser (or runtime) would have to decide if "root.module1.counter1" is a known object (and in that case assign to the objects' value-property) or if it is unknown, it would have to assign directly to "root.module1.counter1.value" (as a scope double-object)
IMHO the parser always needs to check each identifier from left to right ("root", then "root.module1", then "root.module1.counter1", etc) to find if this is a known object (and in that case assign to this c# object) or otherwise (if it reaches the full name "root.module1.counter1.value" without finding a known object) it would assing/add to the double-object in scope

maybe it already works somehow like that? I don't know... but it would be good.
otherwise I fear this left-to-right checking would result in some performance degredation of the engine (if possible at all)

plus: what happens to pre-compiled functions, if - during runtime - someone instantiates an object "root.module1.counter1".
any assignments to "root.module1.counter1.value" _before_ that object instantiation would assign to the double-object in scope; any assignment _after_ the instantiation would assign to the "counter1"-object's value-property.
(that's fine for me, but wouldn't it be necessary to re-compile the term in that case? no idea what happens during compilation... maybe compiling just builds an evaluation-"tree" from the term and still processes/checks each identifier whenever called later on? so in that case it should be OK. however, I could imagine that this might be a problem...)

--> nevertheless: if there is any realistic chance that this enhancement could be part of mages soon, it would make things A LOT easier for me!

PS: I tried using namings like "root»module1»counter1»value=42" and "root»module1»counter2.value=42". 
no ambiguities here! this works perfectly (as the "»" character is part of [#x80 - #xFFFF]) - and it also clearly shows the hierarchy.
but unfortunately, I really need to use "." as the hierarchy seperator (and naturally it's also needed to access the methods and propertys of objects)

**Comment by FlorianRappl** [comment]
> would most likely do the trick for me - however, wouldn't that result in ambiguities? - like:

yes, that's why that is no option. The only option is to make a generic string here as proposed.

**Comment by hanlectin** [comment]
ok - thanks.

I most likely will change the delimiter for our hierarchy after all. Using "\" in the code and replacing it with a character >=0x80
`root\module1\counter1.value=42`
might make things clearer after all and will just require getting used to using '\' instead of '.'
